capnpc/src/codegen.rs

```diff
@@ -1280,15 +1280,15 @@ fn generate_node(gen: &GeneratorContext,
                     reader_members.push(
                         Branch(vec!(
                             Line("#[inline]".to_string()),
-                            Line(format!("pub fn get_{}(self) {} {{", styled_name, ty)),
+                            Line(format!("pub fn get_{}(&self) {} {{", styled_name, ty)),
                             Indent(Box::new(get)),
                             Line("}".to_string()))));
 
                     let (ty_b, get_b, _) = getter_text(gen, &field, false, true)?;
                     builder_members.push(
                         Branch(vec!(
                             Line("#[inline]".to_string()),
-                            Line(format!("pub fn get_{}(self) {} {{", styled_name, ty_b)),
+                            Line(format!("pub fn get_{}(&self) {} {{", styled_name, ty_b)),
```

Comment:
Needs to be `&mut self` to enforce non-aliasing of mutable references.

Reply:
Hmm, I don't quite follow - how is it possible to mutate through the object this returns?

Reply:
Without this `mut`, we could have
```capnp
struct Foo {
  blob @0 : Data;
}
```

```rust
    fn return_aliased<'a>(foo: &'a foo::Builder<'_>) -> capnp::Result<(&'a mut [u8], &'a mut [u8])> {
        Ok((foo.get_blob()?, foo.get_blob()?))
    }
```

I.e. a function that returns two mutable slices pointing to the same memory.

Reply:
Probably `capnp::private::StructBuilder` should *not* derive `Copy`: https://github.com/capnproto/capnproto-rust/blob/2832d59d197ff9c99603e659f02dad0bdd8dfdee/capnp/src/private/layout.rs#L2980-L2981

Then then methods would be forced to take a `&mut self` and it would be less confusing.

Reply:
hm... actually I think that in order for safety to work out we need it to be `&'a mut self` (so that the lifetime is constrained to be the same as the lifetime of the returned `Builder<'a>`). And once we make that change, I think that unfortunately the examples we've been talking about fail to pass the borrow checker.

Reply:
A concrete illustration of what can go wrong if `Builder::get` takes `&mut self` instead of `&'a mut self`:
```rust
    fn return_aliased<'a>(foo: &'a mut foo::Builder<'_>) -> capnp::Result<(&'a mut [u8], &'a mut [u8])> {
        Ok((foo.get_blob()?, foo.get_blob()?))
    }
```

That is, again we find that it's possible to have multiple mutable references to the same memory.